src/kernel/sporadic.c

```diff
@@ -255,20 +255,38 @@ void refill_budget_check(ticks_t usage)
     assert(!isRoundRobin(sc));
     REFILL_SANITY_START(sc);
 
-    /* We assume that the estimates on worst-case kernel overheads are
-     * an upper bound and that overrun cannot occur. */
-#ifdef CONFIG_DEBUG_BUILD
-    if (usage > refill_head(sc)->rAmount) {
-        ticks_t overrun = usage - refill_head(sc)->rAmount;
-        userError(
-            "SC (%p) overran budget by %lluus (%llu ticks)",
-            sc,
-            ticksToUs(overrun),
-            overrun
-        );
+    /*
+     * We charge entire refills in a loop until we end up with a partial
+     * refill or at a point where we can't place refills into the future
+     * without integer overflow.
+     *
+     * Verification actually requires that the current time is at least
+     * 3 * MAX_PERIOD from the INT64_MAX value, so to ease relation to
+     * that assertion we ensure that we never delate a refill past this
+     * point in the future.
+     */
+    while (
+        refill_head(sc)->rAmount <= usage &&
+        INT64_MAX - refill_head(sc)->rTime >= 3 * usToTicks(MAX_BUDGET_US)
```

Comment:
Is `usToTicks(MAX_BUDGET_US)` the same as `UINT64_MAX / 8` at this point?

Reply:
So this condition would stop working once `refill_head(sc)->rTime` reached around 2^61 ticks?

Reply:
How come UINT64_MAX can't be used here instead of INT64_MAX?

Reply:
* `usToTicks(MAX_BUDGET_US)` is `UINT64_MAX` but truncated top fit when converted, so this makes limit smaller
* Yes, we effectively give up on charging budget at around 2^61 ticks
* Should be `UINT64_MAX`

Reply:
Sorry, yes, with the change `usToTicks(MAX_BUDGET_US)` should be effectively `UINT64_MAX / 8` here if no explicit maximum has been configured.